### PR TITLE
Upgrade ConfigCpace to v1.2.0

### PR DESCRIPTION
**Pull request by sarah-segel**
Upgrade ConfigSpace and Numpy
- Exchange deprecated ConfigSpace methods
- Changes to comply with mypi standards for new ConfigSpace and Numpy versions
- Needed to adapt the `get_config_id()` method:
  - So far, iterated through a dictionary of configurations (self.configs) to find the id associated with a given config. If a match was found, returned the corresponding id.
  - As the new ConfigSpace version changed the float rounding when creating a Configuration object, this did no longer work, as configs saved with a previous version would have a different rounding, and just applying the rounding to each config in the comparison would slow things down for runs with many configs.
  - New approach: 
    - Introduce a `config_id_mapping` dict, mapping config tuples of the form (k1, v1, k2, v2, ...) to an id.
    - Apply rounding and add configs to `config_id_mapping` as tuple when adding a new config for a run.
    - To get the id for a given config, apply rounding, convert to tuple and check if the tuple is in `config_id_mapping`. If so, return the id.

### CODE REVIEW

File: deepcave/runs/converters/amltk.py [hunk 1]
```diff
@@ -93,11 +95,9 @@ def from_path(cls, path: Union[Path, str]) -> "AMLTKRun":
         path = Path(path)
 
         # Read configspace
-        from ConfigSpace.read_and_write import json as cs_json
-
         with open(path / "configspace.json", "rb") as f:
             json_string = pickle.load(f)
-        configspace = cs_json.read(json_string)
+        configspace = ConfigurationSpace.from_json(StringIO(json_string))
```
Comment (eddiebergman) [inline deepcave/runs/converters/amltk.py:100]: Interesting that you read it as bytes with `"rb"`.

You could however just do `ConfigurationSpace.from_json(path / "configspace.json")` if your file is actually json (and not byte encoded)

Comment (eddiebergman) [reply deepcave/runs/converters/amltk.py:100]: Oh I see you've done that elsewhere, interesting that it's specifically here

Comment (sarah-segel) [reply deepcave/runs/converters/amltk.py:100]: Thanks for checking!

Yes, the reason for this was that when saving the configspace in AMLTK via
`bucket.store({"configspace.json": cs_json.write(space)})`
I got back an encoded json file. Not sure if that is the correct way to save the json there then though?

Comment (eddiebergman) [reply deepcave/runs/converters/amltk.py:100]: Oh, should have just been `bucket.store({"configspace.json": space.to_serialized_dict()})`, [ref](https://github.com/automl/ConfigSpace/blob/810c8d8a2955f36b91e89a8f2a924a0f74bc6a67/src/ConfigSpace/configuration_space.py#L952) then!

The `cs_json.write(space)` converts it to a binary string, hence the issue

File: deepcave/runs/__init__.py [hunk 2]
```diff
@@ -481,15 +486,17 @@ def get_config_id(self, config: Union[Configuration, Dict]) -> Optional[int]:
...
-        for id, c in self.configs.items():
-            if c == config:
-                return id
-
-        return None
+        # Check if the input configuration tuple exists in the config id mapping
+        if input_config_tuple in self.config_id_mapping:
+            return self.config_id_mapping[input_config_tuple]
+        else:
+            return None
```
Comment (eddiebergman) [inline deepcave/runs/__init__.py:499]: Nice, this should be much faster when you have a lot of configs as it's a constant time lookup :)

File: deepcave/utils/configspace.py [hunk 3]
```diff
@@ -37,16 +38,15 @@ def sample_border_config(configspace: ConfigurationSpace) -> Iterator[Configurat
     while True:
         config = {}
         # Iterates over the hyperparameters to get considered values
-        for hp_name, hp in zip(
-            configspace.get_hyperparameter_names(), configspace.get_hyperparameters()
-        ):
+        for hp_name, hp in zip(list(configspace.keys()), configspace.get_hyperparameters()):
```
Comment (eddiebergman) [inline deepcave/utils/configspace.py:41]: You can do `for hp_name, hp in configspace.items():` :)

File: deepcave/utils/configspace.py [hunk 4]
```diff
@@ -94,16 +94,15 @@ def sample_random_config(
         config_dict = {}
 
         # Iterates over the hyperparameters to get considered values
-        for hp_name, hp in zip(
-            configspace.get_hyperparameter_names(), configspace.get_hyperparameters()
-        ):
+        for hp_name, hp in zip(list(configspace.keys()), configspace.get_hyperparameters()):
```
Comment (eddiebergman) [inline deepcave/utils/configspace.py:97]: Same here

**Review by helegraf — APPROVED** [review]
From the DeepCAVE perspective, this looks fine to me, though I am not an expert on the new ConfigSpace version and how to use it most efficiently

**Review by eddiebergman — COMMENTED** [review]
Just some more small items from ConfigSpace

**Comment by helegraf** [comment]
Though I almost forgot, I am getting some warnings? when I start deepcave

```
The field 'default' should be 'default_value' !
Found in item {'name': 'batch_size', 'log': False, 'lower': 30, 'upper': 300}

/opt/anaconda3/envs/DeepCAVE--upgrade_configspace/lib/python3.9/site-packages/ConfigSpace/configuration_space.py:1052: UserWarning:

The field 'default' should be 'default_value' !
Found in item {'name': 'learning_rate', 'choices': ['constant', 'invscaling', 'adaptive'], 'weights': None}

/opt/anaconda3/envs/DeepCAVE--upgrade_configspace/lib/python3.9/site-packages/ConfigSpace/configuration_space.py:1052: UserWarning:

The field 'default' should be 'default_value' !
Found in item {'name': 'learning_rate_init', 'log': True, 'lower': 0.0001, 'upper': 1.0}

/opt/anaconda3/envs/DeepCAVE--upgrade_configspace/lib/python3.9/site-packages/ConfigSpace/configuration_space.py:1052: UserWarning:

The field 'default' should be 'default_value' !
Found in item {'name': 'alpha', 'log': True, 'lower': 1e-08, 'upper': 1.0}

/opt/anaconda3/envs/DeepCAVE--upgrade_configspace/lib/python3.9/site-packages/ConfigSpace/configuration_space.py:1052: UserWarning:

The field 'default' should be 'default_value' !
Found in item {'name': 'batch_size', 'log': True, 'lower': 4, 'upper': 256}

/opt/anaconda3/envs/DeepCAVE--upgrade_configspace/lib/python3.9/site-packages/ConfigSpace/configuration_space.py:1052: UserWarning:

The field 'default' should be 'default_value' !
Found in item {'name': 'depth', 'log': False, 'lower': 1, 'upper': 3}

/opt/anaconda3/envs/DeepCAVE--upgrade_configspace/lib/python3.9/site-packages/ConfigSpace/configuration_space.py:1052: UserWarning:

The field 'default' should be 'default_value' !
Found in item {'name': 'learning_rate_init', 'log': True, 'lower': 1e-05, 'upper': 1.0}
```

There are more but they're all the same.

**Comment by eddiebergman** [comment]
> From the DeepCAVE perspective, this looks fine to me, though I am not an expert on the new ConfigSpace version and how to use it most efficiently

There's some ergonomic improvements that can be made use of, as well as just exposing some private functions like `_inverse_transform` to a public one like `to_vector`... I never knew which direction _inverse was meant to be.

As for the warnings, that is because `default` was renamed `default_value` (long story as to why that was needed...). Everything serialized from `1.0` would save it as `"default_value"` but we still allow loading old spaces, those with `"default"`.

I should have probably marked these with a `DeprecationWarning` explicitly but you can ignore them where you load it in with: https://docs.python.org/3/library/warnings.html#temporarily-suppressing-warnings

I do not want to remove the warnings though, as I do not want to maintain two possibilities throughout for long term. You can also opt to just update all your config spaces once by loading them in and re-saving them.

**Comment by sarah-segel** [comment]
Thank you for all the hints, I will add them in and exchange the AMLTK examples with the actual json configspaces.

And yes, for the warnings I'd think it makes sense if we update the configspaces for our older example logs once. If users load in their runs created with an older configspace version, it's still fine to show them.

**Comment by sarah-segel** [comment]
@eddiebergman And I agree it would be great to change the warning to a `DeprecationWarning` at some point, as right now it is a bit unclear for users where this warning is coming from (even though we replace our examples, users might still load runs created with the old configspace version and then get the warning).

**Pull request merged**